File: src/ch6/heap_sort.py
import copy
import math


def heap_sort(array):
    array = build_max_heap(array)
    i = len(array) - 1
    output = []
    while i >= 0:
        output.insert(0, array[0])
        array = array[1:]
        array = max_heap(array, 0)
        i = i - 1
    return output


def build_max_heap(array):
    i = math.floor(len(array)/2)
    while i > 0:
        array = max_heap(array, i)
        i = i - 1
    return array


def max_heap(array, loc):
    if loc == 0:
        array = largest_value_first(array)
        max_heap(array, 1)
    left = 2 * loc
    right = 2 * loc + 1
    if left <= len(array)-1 and array[left] > array[loc]:
        largest = left
    else:
        largest = loc
    if right <= len(array)-1 and array[right] > array[largest]:
        largest = right
    if largest != loc:
        array[loc], array[largest] = array[largest], array[loc]
        max_heap(array, largest)
    return array


def largest_value_first(array):
    max_val_loc = 0
    for i in range(1, len(array)-1):
        if array[i] > array[max_val_loc]:
            max_val_loc = i
    if max_val_loc != 0:
        array[0], array[max_val_loc] = array[max_val_loc], array[0]
    return array


if __name__ == '__main__':
    array = [5, 1, 3, 4, 2]
    print(build_max_heap(array))
    print(heap_sort(array))

File: src/ch6/heap_sort.go
package ch6

import "math"

func HeapSort(input []int) (output []int) {
	input = BuildMaxHeap(input)
	for i := len(input) - 1; i >= 1; i-- {
		output = append([]int{input[0]}, output...)
		input = input[1:]
		input = MaxHeap(input, 0)
	}
	return output
}

// BuildMaxHeap creates a max heap complete binary tree
func BuildMaxHeap(input []int) (output []int) {
	halfLength := int(math.Floor(float64(len(input) / 2)))
	for i := halfLength; i >= 0; i-- {
		input = MaxHeap(input, i)
	}
	output = input
	return output
}

// MaxHeap moves the item at the passed in location down to it's appropriate place in the max-heap tree
func MaxHeap(input []int, loc int) (output []int) {
	if loc == 0 {
		MaxHeap(LargestElementFirst(input), 1)
	}
	left := LeftBinaryTreeLocation(loc)
	right := RightBinaryTreeLocation(loc)
	var largest int
	if left <= len(input)-1 && input[left] > input[loc] {
		largest = left
	} else {
		largest = loc
	}
	if right <= len(input)-1 && input[right] > input[largest] {
		largest = right
	}
	if largest != loc {
		input[loc], input[largest] = input[largest], input[loc]
		MaxHeap(input, largest)
	}
	output = input
	return output
}

func LargestElementFirst(input []int) (output []int) {
	maxValLoc := 0
	for i, val := range input {
		if input[maxValLoc] < val {
			maxValLoc = i
		}
	}
	if maxValLoc != 0 {
		input[0], input[maxValLoc] = input[maxValLoc], input[0]
	}
	output = input
	return output
}

func LeftBinaryTreeLocation(loc int) (out int) {
	out = 2 * loc
	return out
}

func RightBinaryTreeLocation(loc int) (out int) {
	out = (2 * loc) + 1
	return out
}
